cpp/src/slice2js/Gen.cpp

```diff
@@ -2570,7 +2593,12 @@ Slice::Gen::TypeScriptVisitor::visitInterfaceDefStart(const InterfaceDefPtr& p)
         _out << nl << fixId(op->name()) << spar;
         for (const auto& param : inParams)
         {
-            _out << (fixId(param->name()) + ":" + typeToTsString(param->type(), true, true, param->optional()));
+            // TypeScript doesn't allow optional parameters with '?' prefix before required parameters.
+            const string optionalPrefix =
+                param->optional() && areRemainingParamsOptional(paramList, param->name()) ? "?" : "";
```

Comment:
The ?: operator is only valid after all required parameters. For optional parameters defined before required parameters, you must explicitly pass undefined.